docs/proposals/20210335-load-balancer-provider.md

```diff
@@ -0,0 +1,999 @@
+    - [Risks and Mitigations](#risks-and-mitigations)
+  - [Alternatives](#alternatives)
+  - [Upgrade Strategy](#upgrade-strategy)
+  - [Additional Details](#additional-details)
+    - [Test Plan [optional]](#test-plan-optional)
+    - [Graduation Criteria [optional]](#graduation-criteria-optional)
+    - [Version Skew Strategy [optional]](#version-skew-strategy-optional)
+  - [Implementation History](#implementation-history)
+
+## Glossary
+
+Refer to the [Cluster API Book Glossary](https://cluster-api.sigs.k8s.io/reference/glossary.html).
+
+If this proposal adds new terms, or defines some, make the changes to the book's glossary when in PR stage.
+
+## Summary
+
+We propose a first-class Load Balancer construct for Cluster API for the purpose of providing a pluggable model for load balancing Kubernetes cluster control planes such that the load balancer implementation isn't tied directly into the Infrastructure Cluster resource. This will open the possibility for multiple swappable options in cloud environments and the ability to re-use load balancer implementations when possible, especially for bare metal and data center environments with heterogeneous infrastructure.
+
+## Motivation
+
+In private data center environments, there is no common infrastructure provider for load balancers vs. the provider for the machines themselves. Virtualized environments may use one vendor for virtual machines, e.g. VMware vSphere, KVM, or Xen, and a variety of vendors for networks and load balancers, e.g. F5, VMware NSX-T, Cisco ACI etc… This requires the ability to plug in different infrastructure providers for networking constructs with that of the machine provisioner. The most common use is the provisioning of a load balancer to sit in front of the Kubernetes control plane instances in order to provide a stable endpoint. Many cluster operators will also want to provision a stable machine-based load balancer to front all workload services from that provisioned cluster.
+
+Additionally, there are also cases where providers such as vSphere may be run in public cloud environments and being able to re-use existing native load balancer implementations would be preferred compared to having to re-implement or fork existing implementations for use in those environments.
+
+There are also cases within existing infrastructure provider implementations where it would be beneficial to allow for easier swapping of different load balancer implementations that may provide different benefits under different circumstances.
+
+We've also seen use cases where there is a need for a Load Balancer to front non-control plane services, such as Ingress. Any functionality that we add should also support these use cases in addition to just control plane load balancing.
+
+### Goals
+
+- Provide a common abstraction for Load Balancers between various providers
+- Allow for reuse of load balancer providers between infrastructure providers where possible
```

Comment:
> I assume the first step when implementing this will be to somewhat extract the existing lb implementations from existing providers. I would assume that providers like CAPO/CAPA/... will still keep the code in the respective repository. Will the extracted lb providers be published via additional release artifacts of CAPO/CAPA/...? (I'm only referring to "default" lb providers which only make sense in combination with a specific infra provider, e.g. OpenStack Octavia)
> 
> I assume we will have two kinds of lb providers. LB providers which only make sense in combination with a specific infra provider and LB provider which are usable with different infra providers. Do we want some kind of compatibility matrix documented somewhere central or should this be included in the documentation of the lb providers? Do we want to make the lb provider easily discoverable? I suppose we just add a section to the cluster-api book?

I do think we'll need to figure out documentation and how to portray the support matrix somehow. As far as discoverability, some of the available options can be exposed through `clusterctl` or the related operator work as built-in providers, similar to how existing infrastructure providers are exposed today. 

> 
> I think from a test perspective it would make sense to leave "infra provider" (CAPO/CAPA) specific lb providers like OpenStack Ocatvia in the respective "infra provider" repositories. Similar like CAPD in the main cluster-api repo.

Indeed, I do believe that it'll make sense to keep provider specific implementations in the same repositories as their current functionality exists, but how they exist there may need to change to facilitate their use.

For example, where there are situations such as a "provider" deployed on top of another "provider", think vSphere (or even OpenStack) on top of a public cloud and leveraging a more native LB solution to the public cloud provider. While not all combinations can or should be supported I know there is a desire to support this on the CAPV side.